working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 400
+      assert response_body == {"message": "description is required"}
+  ```
+</details>
+
+Now that we've seen several facets of working with Copilot to write tests, we will leave it as a personal exercise to complete the test suite for our last untested dependency `delete_book`.
+
+#### Executing the Refactor
+
+Now that we have a solid test suite, we can get some help from Copilot to make this refactor go a little quicker. Let's review what needs to change in `validate_book` to make the function more flexible so we can use it with other models:
+- take in a second parameter that represents a model class reference
+- change any uses of the `Book` class to use our new parameter
+- update the function name, variable names, and return messages so they reflect that the function is not specific to the `Book` class
+
+The changes we need are pretty minimal and could be quick to perform manually, but for the practice, let's see how Copilot handles the change. If we highlight the `validate_book` function, open up the inline Copilot chat, and use the prompt:
+
+> Please update this function so that we could use it for any model class
+
+The response delivers exactly what we're looking for:
+```
+def validate_model(model_class, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} invalid"}, 400))
+
+    model = model_class.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} not found"}, 404))
+
+    return model
+```
+
+Our function is updated, now we just need to update our dependent functions and tests to use `validate_model` instead of `validate_book`, which includes passing in the new `model_class` parameter. We can do this manually using VS Code's Find & Replace tools then adding in our new parameter, or we can reach out to Copilot. 
+
+One option is to open each file, select the whole file, and ask Copilot to change any instances of `validate_book` to `validate_model`. We need to carefully review the lines Copilot suggests changing to ensure nothing unexpected is altered, but it can save us a little time since we need to do slightly more than find and replace the function name.
+
+In our experimentation, we found that in the `routes.py` file, we could use a prompt with very few details to get the changes that we wanted: 
+
+> Please replace all uses of validate_book with validate_model
+
+However, in the `test_routes.py` file, we had to add more detail for Copilot to recognize that there was a new class parameter which was required wherever the function was invoked, and that we needed import the `Book` class:
+
+> Please replace all uses of validate_book(book_id) with validate_model(model_class, model_id). Import any required model classes.
```

Comment:
- rephrased to make it sound less like "we" are talking at the reader. we prefer to use "we" inclusively, but here, "we" refers to the person who wrote this who is reporting their results.
```suggestion
Due to differences in context, it's possible that a prompt that works fine in one file may not work as well in another file. For example, in the `routes.py` file, the following prompt, which provides very few details, is able to produce the desired changes:

> Please replace all uses of validate_book with validate_model

However, in the `test_routes.py` file, more detail is necessary for Copilot to recognize that there is a new class parameter which is required wherever the function is invoked, and that the `Book` class needs to be imported:

> Please replace all uses of validate_book(book_id) with validate_model(cls, model_id). Import any required model classes.
```